### ISSUE TITLE
possible bug in "using geospatial data"

### ISSUE THREAD

**Issue by mdlincoln** [programminghistorian/jekyll#1181]
[reported from twitter](https://twitter.com/Nate_Housley/status/1097272622834798598):

>There seems to be a directory missing in the data provided for "Using Geospatial Data to Inform Historical Research in R." The tutorial reads: Census_Data <- read.csv("./data/County1990_Data/nhgis0027_ts_nominal_county.csv" but there is no County1990_Data folder

I can look into it this weekend unless someone has a chance to take a look sooner.

**Comment by mdlincoln** [comment]
OK, I've checked through the lesson and the uploaded data have quite a few problems.

Firstly, our reporter is correct - there is no `Country1900_Data/nhgist0027_ts_nominal_country.csv` file.

Moreover, the `County1900` directory data is incomplete - it has data, but no spatial information, such that when I run the very first import command of the lesson, I get the following warning:

```r
County_Aggregate_Data <- st_read("./data/County1990/")
#> Reading layer `US_county_1990' from data source `/Users/mlincoln/Development/geospatial/data/County1990' using driver `ESRI Shapefile'
#> Warning message:
#> no simple feature geometries present: returning a data.frame or tbl_df 
```

Because of this, none of the rest of the lesson will work. FWIW I also tried opening this shapefile in QGIS (which also ought to be able to read it) and it says that it's not a valid shapefile, so whatever data have been uploaded are corrupt or incomplete.

@JMParr you were the editor for this lesson, can you please check if you have the complete data for it somewhere?

**Comment by JMParr** [comment]
Hi, @mdlincoln - I've looked through the files, including old clones from Github. Unfortunately, it looks like I do not have complete data. I'm not sure what happened. Correcting it will necessitate contacting the author.

**Comment by mdlincoln** [comment]
@JMParr ok, please contact them. Without it the lesson is unusable.

**Comment by acrymble** [comment]
I'm going to put a warning message on this until this can be resolved. We don't want people thinking this lesson is useable.

**Comment by acrymble** [comment]
@ladew222 you were the author of this tutorial. It looks like there has been an issue getting all of the data files in place. Can you help us rectify this so we can re-publicize the lesson?

**Comment by mdlincoln** [comment]
Thanks for adding that warning @acrymble, I think that's a reasonable interim action to take. @JMParr have you emailed the author yet? 

If we aren't able to get the correct data file within some period of time (1-2 months?), though, I think it would be necessary to [retire the lesson](https://programminghistorian.org/en/lesson-retirement-policy) This isn't a minor issue in one section - without the correct data almost all of the lesson is broken.

**Comment by JMParr** [comment]
I have emailed the author, but I haven't heard back. It's possible it went
into the spam folder, or he's just busy and hasn't had a chance to reply.
The data files were fine when the lesson was originally finalized.

**Comment by ladew222** [comment]
I apologize for the late reply as there was an email issue.  I will take a look and get back to you this week. Like Jessica mentioned, the files were functioning at publication so the files may have gotten corrupted somehow if I was to guess. I will look to see what is the source of the issue and repackage the files if necessary.

**Comment by JMParr** [comment]
Thank you @ladew22 !

**Comment by ladew222** [comment]
No problem. I compared my archive to the compressed download and there is quite a bit missing; both files and folders. I checked to see very briefly to see the code worked with my archive and it appears to function but I want work through the lesson more completely. Once done, I can send over the files.  What is the best means for me to deliver the items.

**Comment by JMParr** [comment]
Sounds good. @acrymble is Managing Editor. He can let us know.

**Comment by acrymble** [comment]
Thank you both for the update. @ladew222 can you try to get us the fix by 7 May 2019 please? You can email them to the managing editor (see http://programminghistorian.org/en/project-team)

**Comment by ladew222** [comment]
I am a day tardy on this due to illness.  I am sending the files today.

**Comment by JMParr** [comment]
Thank you @ladew222 . I hope you are feeling better. Tagging @mdlincoln in case he hasn't seen this yet.

**Comment by acrymble** [comment]
I've received the data from @ladew222 and he says there are 2 other minor fixes he wants to make. Once I hear what those are I'll do a pull request with the changes.

**Comment by acrymble** [comment]
I've received the data and the author has suggested the following two fixes:

County_Aggregate_Data$RelativeTotal= ((cntyNCG$AV0AA1990/10000)/cntyNCG$CountMembers )

To

County_Aggregate_Data$RelativeTotal= ((County_Aggregate_Data$AV0AA1990/10000)/County_Aggregate_Data$CountMembers )



And

qtm(shp = cntyNCG, fill = "CountMembers",text="NHGISNAM",text.size="A57AA1980")

to

qtm(shp = County_Aggregate_Data, fill = "CountMembers",text="NHGISNAM",text.size="A57AA1980")


--

I will push up the changes and then we can make sure everything is still working.

**Comment by acrymble** [comment]
We have a new problem.

The shapefiles for this lesson are too large. The file "County1990/US_county_1990.shp" is 79mb, and as a result, the Archive.zip file that contains everything is also very large. Github will only let me upload files of 25mb. That means that even with these fixes a key file is going to be missing.

@mdlincoln is there a solution we can adopt? I've half-completed these changes on branch "issue-1181".

I have the missing files on my computer

**Comment by ladew222** [comment]
I will try to reduce the file and try to limit it to the two states I directly analyzed rather than the entire US.  I will try to reduce on my end with hat 25mb number as a reference.

**Comment by mdlincoln** [comment]
Thanks @ladew222 - yes, please simplify the geometry as this lesson doesn't require anything near the resolution of 79mb. https://mapshaper.org/ is a very simple online tool for doing this, or there are methods in R, QGIS, ArcMap for it as well.

**Comment by ladew222** [comment]
Thanks,

I used map shaper and simplified the shape file.  The files is now 23.2 mb.  I can reduce more if needed as well. Here is a link to the modified file: https://we.tl/t-bZ50NVAjwj <https://we.tl/t-bZ50NVAjwj>.

**Comment by acrymble** [comment]
I've got the smaller file. I'm going to try to do a pull request. @ladew222 I think ideally a much smaller file would make the lesson more globally accessible. We've been reminded recently that some parts of the world still struggle to download files of this size. But this resolves our immediate need.

**Comment by acrymble** [comment]
Oops @ladew222 I've just noticed that your new file, when extracted from the zip, is still above 25mb. We've uploaded the files individually into the assets folder, so this too needs to be below the threshold. I haven't opened the file itself to see what extra detail is in there, but is there more that can be cut that allows for a minimally functioning tutorial with small file sizes?

**Comment by ladew222** [comment]
Sure. I will cut a bit more and make is well under to prevent any issues.

**Comment by ladew222** [comment]
I reduced the file down further. I apologize for the too large file. I uploaded the entire County1990 folder which is now listed at 16.9 mb.   Here is the file https://we.tl/t-eumryd43pt <https://we.tl/t-eumryd43pt>



https://we.tl/t-Y9buBfVS6s <https://we.tl/t-Y9buBfVS6s>

**Comment by acrymble** [comment]
I've created a pull request where we can preview this lesson.

https://deploy-preview-1312--ph-dev.netlify.com/en/lessons/geospatial-data-analysis

Is anyone available to try out the code to help us make sure it works now?

**Comment by ladew222** [comment]
Thanks for anyone willing to help. I will run through the complete lesson again myself as well over the next couple of days.

**Comment by acrymble** [comment]
@ladew222 I'm afraid we're still having problems with this, which @mdlincoln has pointed out in his review of the pull request. https://github.com/programminghistorian/jekyll/pull/1312

Can you please address the comments, and completely re-try the lesson before we try another pull request?

If we aren't able to fix this by the end of June, I'm going to propose retiring the lesson.

**Comment by ladew222** [comment]
Thanks everyone. I noted similar issues yesterday and was trying alternative toward simplification of the geometry to deal with the size issues but I think @mdlincoln <https://github.com/mdlincoln> may be right that the best route is to subset North Carolina and South Carolina rather that simplifying the geometry as there appears to have been an issue with map shaper’s simplification. I do cover subsetting as part of the lesson so  I may leave in one other state so I can describe that process as long as file size is not an issue. I do have an addendum that address how to obtain the entire shapefile from NHGIS so it should cover users who are interested in other geographic regions at least in the U.S. I will then test the entire lesson with the new files and update you when it is tested and complete.

**Comment by mdlincoln** [comment]
@ladew222 I like that idea a lot!

**Comment by ladew222** [comment]
I was able to simplify the entire U.S. map and test the results which saved me from some additional modifications to the actual lesson.  I went through the lesson and tested the new files to ensure they work properly. The data directory is available here:https://dropmefiles.com/0OP4Q <https://dropmefiles.com/0OP4Q> and is 32 MB.  The .shp file is no 19.2 MB and is the largest file. I updated the lesson file from ph-submissions at https://github.com/programminghistorian/ph-submissions/blob/gh-pages/lessons/geospatial-data-analysis.md <https://github.com/programminghistorian/ph-submissions/blob/gh-pages/lessons/geospatial-data-analysis.md>.  This is needed to match the new data. The most substantive change is that the data directory is now County1990ussm.  If you are able to upload the files, I am more than willing to test again once uploaded or make any additional changes. 

Thanks everyone.

**Comment by acrymble** [comment]
I believe @JMParr is back in a few days. Can you please take over and re-test this lesson so that we can resolve the issues with the data files?

Let us know if you need help.

**Comment by acrymble** [comment]
We had suggested retiring this lesson if a fix was not in place by the end of June. @JMParr we're now  at that deadline. Can you please let us know your plan so we can proceed appropriately?

**Comment by ladew222** [comment]
And please let me know if anything is needed from my end.

**Comment by JMParr** [comment]
Hi, Adam- I'm just back from medical leave, and was away last week. I will
take a look through this tomorrow. Sorry for the delay. I did not see your
previous email.

**Comment by JMParr** [comment]
Just a confirmation that I am working on testing. I'll provide an update by the end of the week.

**Comment by JMParr** [comment]
Hi, all. Thank you for your patience. I've hit a bump that's probably minor, and I wanted to double check that it was not an issue local to my fairly elderly laptop. That meant getting into campus to try on some other machines (and you can imagine the traffic getting near a Boston campus the week of July 4th).  In trying to run the package installs, I keep getting requests to select HTTPS CRAN mirror.

**Comment by JMParr** [comment]
@ladew222 is this a step that comes up on your end?

**Comment by mdlincoln** [comment]
@JMParr If someone has installed R for the first time, then R will prompt them to select a CRAN mirror from the list provided. Once you do that, it ought to continue installation with no problems - were you able to do so?

**Comment by JMParr** [comment]
I've had R installed on my laptop for 3 years, and had an updated version. Similar issue on campus, and none of them were new installs. Very strange.

**Comment by mdlincoln** [comment]
@JMParr the setting gets stored in an `.Rprofile` file in your home directory, so it's possible that got moved somehow? In any event, nothing in the lesson code would trigger that.

Also - AFTER you selected that, were you able to proceed?

**Comment by JMParr** [comment]
Thanks, @mdlincoln . My R skills are admittedly rusty. New error message, asserting that there's no package called "tmap." Here's the full text of the error: 

The downloaded source packages are in
	‘/private/var/folders/bb/yy26r_g54_136qgvyj_3c_nm0000gn/T/RtmpFKKvA6/downloaded_packages’
Warning message:
In install.packages("sf") :
  installation of package ‘units’ had non-zero exit status
> library(sf)
configure: error: in `/private/var/folders/bb/yy26r_g54_136qgvyj_3c_nm0000gn/T/RtmpqWzSF2/R.INSTALL814e200d79/units':
configure: error: C++ compiler cannot create executables
See `config.log' for more details
ERROR: configuration failed for package ‘units’
* removing ‘/Library/Frameworks/R.framework/Versions/3.3/Resources/library/units’
Error in loadNamespace(j <- i[[1L]], c(lib.loc, .libPaths()), versionCheck = vI[[j]]) : 
  there is no package called ‘units’
In addition: Warning message:
package ‘sf’ was built under R version 3.3.2 
Error: package or namespace load failed for ‘sf’
> library(tmap)
Error in library(tmap) : there is no package called ‘tmap’
>

**Comment by ladew222** [comment]
I will do a raw install on a new machine here today to see if I can replicate any of the issues with the packages in R.  I have not had issues but have had R installed for some time on both my machines.

**Comment by JMParr** [comment]
Thanks @ladew222 - I'm also going do a new one to see if that helps. It's possible that something has gone wrong with my install through an update or something else.

**Comment by JMParr** [comment]
Hmmm. That didn't help. Very strange.

**Comment by mdlincoln** [comment]
ooh, `units` often causes issues because it hooks into a lot of C++ code. First off though, it looks like you're running a very old version of R - the current version is 3.6.1 and it looks like yours is around 3.3? I'm afraid that's the most troubleshooting I can do today, but it may be helpful to just wipe out everything and rebuild from the newest version.

**Comment by JMParr** [comment]
Thanks! I thought I had recently updated, but it looks like I was mistaken. Am working on another new wipe and re-install, but need to update my OS first.

**Comment by mdlincoln** [comment]
good luck! I had to do all that last month, myself... 😭

**Comment by ladew222** [comment]
Yep. I will try a raw here once I get access to another machine to see how that runs as well.

**Comment by JMParr** [comment]
Heh. Thanks! I may have been hedging my bets. New OS update for my laptop is quite recent, but laptop is now 4 years old, "well-loved," and showing its age. Was looking to replace it this summer, but will just have to take the plunge and run those updates now since it's become a productivity/work issue.

**Comment by ladew222** [comment]
I was able to do a raw install here without problems.  Please let me know if I can be of assistance on the checking/update.

**Comment by JMParr** [comment]
Looks good on my end as well. I think we can go ahead and update.

**Comment by mdlincoln** [comment]
ok, one of you then please start a PR

**Comment by acrymble** [comment]
@alsalin can you, as ME, make sure this is resolved?

**Comment by ladew222** [comment]
I have not done this before, but if I can help or "sign off" on it, I am more than willing to do that.

**Comment by rivaquiroga** [comment]
Hi! What is the status of this issue? Any idea when it would be solved? There is someone interested in translating this lesson to Spanish. It would be great if I can tell her when she would be able to start, to not lose the momentum. 
If there is anything I can do to help, please let me know :nerd_face: 
Thanks!

**Comment by ladew222** [comment]
Thanks River. I am glad to see the interest.  Let me know if I can be of service in the process and let me know if I can help getting this pull request finalized.

**Comment by JMParr** [comment]
Hi, Riva- I'm in the process of moving the files. I hope to finish next
week. I've been delayed because our semester just started, and I was asked
to take on a course last minute.  I'll be in touch soon.

**Comment by acrymble** [comment]
This remains stuck. Just to clarify, the process for uploading these files is in: https://github.com/programminghistorian/jekyll/wiki/Making-Technical-Contributions

**Comment by JMParr** [comment]
It's in progress. I've been working 90+ hour weeks and have not yet had a
chance to finish working on it.

**Comment by acrymble** [comment]
Sure @JMParr we certainly understand. If it helps you to get organized when you do have time, it might be worth posting a list of the files involved that will need to be uploaded. Then it is just a matter of doing the upload when you've got a moment.

We'll have a new managing editor soon who might be able to help with the upload as a way of  practicing the process. @ladew222 if you know the list of files affected, please post them here so we've got them ready.

**Comment by JMParr** [comment]
Thanks, Adam! In addition to updating the YAML, I do need to finish making
a list of the files so that I can insure that everything has been moved
during the process. If the new ME is able to get to the upload before I
have a chance, that would be great. I know this has been sitting for a
while and needs to be finished.

**Comment by acrymble** [comment]
@alsalin @svmelton I think @JMParr could use some help resolving this issue with a broken lesson in the English publication.

**Comment by svmelton** [comment]
@alsalin and I are going to talk through this and will check back.

**Comment by ladew222** [comment]
Let me know if you have questions from me.

**Comment by JMParr** [comment]
> @alsalin and I are going to talk through this and will check back.

Hi, @svmelton - This revised lesson needs to be published. I've been absolutely buried in teaching and service this semester. It was previously published, but something happened post-publication, and it stopped working.  @ladew222 finished fixing it, and it's ready to go. It's just a complex move because there are a fair number of files attached to the lesson.

**Comment by svmelton** [comment]
Thanks for the update, @JMParr! Is there anything we can do to help?

**Comment by mdlincoln** [comment]
@svmelton @JMParr I was hoping to let this one resolve itself without needing direct input by the tech team, because this should be a routine editorial task. but I also recognize the complexity of this particular fix, and that we've had so many switches of the ME. 

Please loop me in and let's get these files into a PR soon. @ladew222 has had to wait long enough for us to facilitate this.

**Comment by svmelton** [comment]
Thanks so much for your patience, @ladew222. I'm going to take over the upload from here; @ladew222, would it be possible for you to send the updated files again? (It looks like the old link expired.) Thanks—I appreciate your help!

**Comment by ladew222** [comment]
No problem.  I will should be able to send them either later today when back on my computer or soon after.

**Comment by ladew222** [comment]
I uploaded the files. Here is the link: https://we.tl/t-fEcZndRzfi <https://we.tl/t-fEcZndRzfi> . I am more than happy to go through the lesson as it has been a while since I have worked through it and I want to make sure the files match the updated lesson. Let me know as well if I can help in any other.

**Comment by svmelton** [comment]
Thanks, @ladew222! I'm going to work on this and will aim to get the PR done by the end of this weekend.

**Comment by ladew222** [comment]
Thanks Sarah.  Let me know if I can be of service.

**Comment by svmelton** [comment]
Just a quick update: it'll be a few more days as I fix the links in the markdown.

**Comment by JMParr** [comment]
Thanks for taking this on, Sarah!

**Comment by ladew222** [comment]
No problem. Thanks.

**Comment by svmelton** [comment]
With the caveat that I'm a bit of a novice in R:
I went through the lesson to make sure everything was working as expected. I ran across the following errors:
1. When trying to run
`plot(points[which(point$State %in% c("NC","SC")),]$geometry,col = "green", pch=20,cex=.5, axes=TRUE,add=TRUE)`
I got an error message of `Error in point$State %in% c("NC", "SC") : object 'point' not found`

2. When running `County_Aggregate_Data$STATEFP <- as.numeric(as.character(County_Aggregate_Data$STATEFP))` and `County_Aggregate_Data$COUNTYFP <- as.numeric(as.character(County_Aggregate_Data$COUNTYFP))`
I receive this error
`Error in `[[<-.data.frame`(`*tmp*`, i, value = numeric(0)) : 
  replacement has 0 rows, data has 146`

At that point, I stopped because the rest of the lesson is dependent on that variable. I'm not sure if the issue is user error or something with the new data. I'd appreciate @ladew222 or anyone from @programminghistorian/technical-team or elsewhere taking a look!

(I'm on R 3.6.1, btw.)

**Comment by ladew222** [comment]
Thanks Sarah. I will take and get back to you tomorrow.

**Comment by rivaquiroga** [comment]
I can also take a look if you want. @ladew222, it is possible for you to share again the link with the files? The previous one is not active anymore

**Comment by svmelton** [comment]
Hi all—after incorporating the changes @ladew222 made [here](https://github.com/programminghistorian/ph-submissions/commit/27eca3618ef1ada7e1bfbfc9cd385d6a5e61b3a8), I'm still running into the same issues. I've [copied the data here](https://www.dropbox.com/sh/ddrme45m36oqo1i/AAA2L0FQOcLjNfTlpaphRiSda?dl=0), and it would be great if anyone had time to take a look.

Thanks!

**Comment by rivaquiroga** [comment]
Hi @svmelton! 
The first one is a typo. You wrote `plot(points[which(point$State...`; it should be `points$State`, because the name of the object is `points`.
I went through the whole code and I'm not getting the second error. I tried to replicate it and I suspect that maybe you forgot to run this line `County_Aggregate_Data <- merge(County_Aggregate_Data,Census_Data,by.x="GISJOIN", by.y="GISJOIN")`. By not running it is the only way I can get the `[[<-.data.frame(tmp, i, value = numeric(0))...` error.

Something else: in the data you share through Dropbox the csv `nhgis0027_ts_nominal_county.csv` is on `
County1990_Data`, not in `
County1990ussm`, so [this change](https://github.com/programminghistorian/ph-submissions/commit/27eca3618ef1ada7e1bfbfc9cd385d6a5e61b3a8#diff-54dec73487d95f701b0daff52e7a404fR90) is not necessary.
And [here](https://github.com/programminghistorian/ph-submissions/commit/27eca3618ef1ada7e1bfbfc9cd385d6a5e61b3a8#diff-54dec73487d95f701b0daff52e7a404fR103), `Addresses` should be capitalized to match the name of the previous object.
Let me know if I can help in anything else :nerd_face:

**Comment by svmelton** [comment]
Thanks so much, @rivaquiroga! That seems to have fixed those issues.

In the last section of the lesson (Visualizing Data Relationships), I'm running into an issue. When I try to run `WhitePer10K <- ((dataM$B18AA1990/dataM$TOTPOP)*10000)`, I get an error of `Error: object 'dataM' not found`.

**Comment by jenniferisasi** [comment]
I was going to try to help but... it seems like the object `dataM` is not called or mentioned in the lesson prior to that section, so you probably didn't load it, right? It is supposed to be the dataset and until that point in the lesson, the data set looks to be `County_Aggregate_Data`. Can't find any other reference to dataM in the downloaded files or the rest of the lesson. 

I haven't done all the lesson but, try to switch `dataM` with `County_Aggregate_Data` as above it looks like you merged all the data into it and it contains all the information generated during the previous sections. Hope it works

**Comment by svmelton** [comment]
Perfect—thank you, @jenniferisasi!

**Comment by svmelton** [comment]
Ok, last issue I'm seeing (this is probably a question for @ladew222):
in `plot(MembersPer10K,ChurchesPer10K)` and `plot(MembersPer10K, log(ChurchesPer10K+1))`, it looks like neither MembersPer10K or ChurchesPer10K has been defined.

I think we'll be ready to update the lesson after that!

**Comment by ladew222** [comment]
Sarah, 

Early on in the update process, I updated the file in phsubmissions at https://github.com/programminghistorian/ph-submissions/blob/gh-pages/lessons/geospatial-data-analysis.md <https://github.com/programminghistorian/ph-submissions/blob/gh-pages/lessons/geospatial-data-analysis.md> to fix some problems in the text after I looked over the file after the data problem was discovered. That was a while back and I suspect that this issue stems from the file https://programminghistorian.org/en/lessons/geospatial-data-analysis <https://programminghistorian.org/en/lessons/geospatial-data-analysis> not reflecting those changes. I should have updated you on that when you started in retrospect. I apologize for the oversight, as I have been testing from that version.  But if we only have this last remaining issue, feel free to bring those changes over—or use the my updated text if easier. I will paste the relevant section form submissions to resolve this variables.  

Below, let's set up a variable to try to take a look at some of the variables to look for possible correlations. Below we are going to create a variable that measures the distribution of denominational churches in a county, which will allow us measure if our membership is correlated with a particular denomination:

Churches_Per10K <- ((County_Aggregate_Data$AOG.C/County_Aggregate_Data$CHTOTAL)*10000)
MembersPer10K <- as.integer(((County_Aggregate_Data$CountMembers/County_Aggregate_Data$TOTPOP)*100000))
And
